flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/storage/SortBufferAccumulator.java

```diff
@@ -0,0 +1,321 @@
+ * thread.
+ */
+public class SortBufferAccumulator implements BufferAccumulator {
+
+    /** The number of the subpartitions. */
+    private final int numSubpartitions;
+
+    /** The total number of the buffers used by the {@link SortBufferAccumulator}. */
+    private final int numBuffers;
+
+    /** The byte size of one single buffer. */
+    private final int bufferSizeBytes;
+
+    /** The empty buffers without storing data. */
+    private final LinkedList<MemorySegment> freeSegments = new LinkedList<>();
+
+    /** The memory manager of the tiered storage. */
+    private final TieredStorageMemoryManager storeMemoryManager;
+
+    /** The number of buffers for sorting used in the {@link SortBufferContainer}. */
+    private int numBuffersForSort;
+
+    /**
+     * The {@link SortBufferContainer} for accumulating broadcast data. Note that this can be null
+     * before using it to store records, and this buffer container will be released once flushed.
+     */
+    @Nullable private SortBufferContainer broadcastDataBuffer;
+
+    /**
+     * The {@link SortBufferContainer} for accumulating non-broadcast data. Note that this can be
+     * null before using it to store records, and this buffer container will be released once
+     * flushed.
+     */
+    @Nullable private SortBufferContainer unicastDataBuffer;
+
+    /**
+     * The buffer recycler. Note that this can be null before requesting buffers from the memory
+     * manager.
+     */
+    @Nullable private BufferRecycler bufferRecycler;
+
+    /**
+     * The {@link SortBufferAccumulator}'s accumulated buffer flusher is not prepared during
+     * construction, requiring the field to be initialized during setup. Therefore, it is necessary
+     * to verify whether this field is null before using it.
+     */
+    @Nullable
+    private BiConsumer<TieredStorageSubpartitionId, List<Buffer>> accumulatedBufferFlusher;
+
+    public SortBufferAccumulator(
+            int numSubpartitions,
+            int numBuffers,
+            int bufferSizeBytes,
+            TieredStorageMemoryManager storeMemoryManager) {
+        this.numSubpartitions = numSubpartitions;
+        this.bufferSizeBytes = bufferSizeBytes;
+        this.numBuffers = numBuffers;
+        this.storeMemoryManager = storeMemoryManager;
+    }
+
+    @Override
+    public void setup(BiConsumer<TieredStorageSubpartitionId, List<Buffer>> bufferFlusher) {
+        this.accumulatedBufferFlusher = bufferFlusher;
+    }
+
+    @Override
+    public void receive(
+            ByteBuffer record,
+            TieredStorageSubpartitionId subpartitionId,
+            Buffer.DataType dataType,
+            boolean isBroadcast)
+            throws IOException {
+        int targetSubpartition = subpartitionId.getSubpartitionId();
+        SortBufferContainer sortBufferContainer =
+                isBroadcast ? getBroadcastDataBuffer() : getUnicastDataBuffer();
```

Comment:
It is implicit that getting one of the two data buffers would result in the other one being flushed.

I'd suggest the following:
- Have only one field `currentDataBuffer`, instead of two nullable `broadcastDataBuffer` and `unicastDataBuffer` and only one of them can have non-null value at anytime.
- Instead of calling `getBroadcast/UnicastDataBuffer`, we can have a method `switchCurrentDataBufferIfNeeded(isBroadcast)`, in which we can update the value of `currentDataBuffer`. This would make the flushing less surprising.
- We might need another field to remember whether the current data buffer is broadcast or unicast.

Reply:
Fixed with a `currentDataBuffer `, a new `switchCurrentDataBufferIfNeeded(isBroadcast)`, a field of `isBroadcastDataBuffer`.